### PR TITLE
Add validation for upstreamNameserver port

### LINKED ISSUE #178: upstreamNameservers does not allow custom port number
While debugging #169 I found out that `upstreamNameservers` in `kube-dns` configmap does not allow port numbers for upstream servers.

Port numbers are accepted if you give them as `--nameserver=server:port` argument for kube-dns container, but it would be nice to have them accepted via configmap as well.

### PR DESCRIPTION
**Pull request by rramkumar1**
Fixes #178. Add validation for a port number for upstream nameservers specified in the config map.

### CODE REVIEW

File: pkg/dns/config/config.go [hunk 1]
```diff
@@ -117,7 +117,15 @@ func (config *Config) validateUpstreamNameserver() error {
 	}
 
 	for _, ns := range config.UpstreamNameservers {
-		if len(validation.IsValidIP(ns)) > 0 {
+		// TODO(rramkumar): Cleanly split this into seperate function.
+		nsStrings := strings.SplitN(ns, ":", 2)
+		// Validate port if specified
+		if len(nsStrings) == 2 {
+			if _, err := strconv.ParseUint(nsStrings[1], 10, 16); err != nil {
+				return fmt.Errorf("Invalid nameserver: %q", ns)
```
Comment (m1093782566) [inline pkg/dns/config/config.go]: nit: first world of error message should be low case.

Comment (rramkumar1) [reply pkg/dns/config/config.go]: I'm not following... what do should the error message be?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [kubernetes/dns] Add validation..." ]
>rramkumar1 commented on this pull request.
>
>
>
>> @@ -117,7 +117,15 @@ func (config *Config) validateUpstreamNameserver() error {
> 	}
>
> 	for _, ns := range config.UpstreamNameservers {
>-		if len(validation.IsValidIP(ns)) > 0 {
>+		// TODO(rramkumar): Cleanly split this into seperate function.
>+		nsStrings := strings.SplitN(ns, ":", 2)
>+		// Validate port if specified
>+		if len(nsStrings) == 2 {
>+			if _, err := strconv.ParseUint(nsStrings[1], 10, 16); err != nil {
>+				return fmt.Errorf("Invalid nameserver: %q", ns)
>
>I'm not following... what do should the error message be?

~~~
return fmt.Errorf("invalid nameserver: %q", ns)
~~~

Why is net.SplitHostPort not used btw?

**Comment by rramkumar1** [comment]
@miekg I initially thought about SplitHostPort but then I figured it might be cleaner to reuse the validation logic from validateStubDomains(). If you prefer SplitHostPort, I can do that and reimplement the validation in validateStubDomains() to use it.

File: pkg/dns/config/config.go (hunk 1 above)
Comment (m1093782566) [reply pkg/dns/config/config.go]: fmt.Errorf("invalid nameserver: %q", ns)

glog.Errorf("Invalid ...")

Comment (bowei) [reply pkg/dns/config/config.go]: Generally speaking, if the purpose of a function is to validate something and return an error, you should not need to log as the error will be logged by the caller up on the call stack. This also avoids spuriously error logging when testing potential configurations.

**Comment by rramkumar1** [comment]
@m1093782566 For some reason when I squashed the commits the comment history went away but I addressed the nit. 

@miekg Tried to use SplitHostPort but it errors when given an IP with no port. This is rough because it means there has to be some hacky logic (i.e find last colon, get port manually). I added a TODO to figure out how to do this cleanly. In the meantime, I would at least like to get this in.

File: pkg/dns/config/config.go (hunk 1 above)
Comment (rramkumar1) [reply pkg/dns/config/config.go]: done

**Pull request merged**